bumble/drivers/intel.py

```diff
@@ -0,0 +1,343 @@
+logger = logging.getLogger(__name__)
+
+# -----------------------------------------------------------------------------
+# Constants
+# -----------------------------------------------------------------------------
+CSS_HEADER_OFFSET = 8
+ECDSA_OFFSET = 644
+OPERATIONAL_FW = 0x03
+INTEL_VENDOR_ID = 0x0000
+HW_PLATFORM = 0x37
+MAX_FRAGMENT_PAYLOAD = 252
+RSA_HEADER_LEN = 644
+ECDSA_HEADER_LEN = 320
+
+LINUX_FW_PATH = "/lib/firmware/intel/"
+
+# -----------------------------------------------------------------------------
+# HCI Commands
+# -----------------------------------------------------------------------------
+HCI_INTEL_READ_VERSION_COMMAND = hci_vendor_command_op_code(0xFC05)  # type: ignore
+HCI_INTEL_SECURE_SEND_COMMAND = hci_vendor_command_op_code(0xFC09)  # type: ignore
+HCI_INTEL_RESET_COMMAND = hci_vendor_command_op_code(0xFC01)  # type: ignore
+HCI_Command.register_commands(globals())
+
+
+@HCI_Command.command(  # type: ignore
+    fields=[("param", 1)],
+    return_parameters_fields=[
+        ("status", STATUS_SPEC),
+        ("data", "*"),
+    ],
+)
+class HCI_Intel_Read_Version_Command(HCI_Command):
+    pass
+
+
+@HCI_Command.command(  # type: ignore
+    # fields=[("param", MAX_FRAGMENT_PAYLOAD + 1)],
+    fields=[("param", "*")],
+    return_parameters_fields=[
+        ("data", "*"),
+    ],
+)
+class Hci_Intel_Secure_Send_Command(HCI_Command):
+    pass
+
+
+@HCI_Command.command(  # type: ignore
+    fields=[
+        ("reset_type", 1),
+        ("patch_enable", 1),
+        ("ddc_reload", 1),
+        ("boot_option", 1),
+        ("boot_param", 4),
+    ],
+    return_parameters_fields=[
+        ("data", "*"),
+    ],
+)
+class Hci_Intel_Reset_Command(HCI_Command):
+    pass
+
+
+# -----------------------------------------------------------------------------
+
+
+@dataclass
+class IntelTLV:
+    type: int
+    lenght: int
+    value: bytearray
+
+
+@dataclass
+class IntelVersionTlv:
+    cnvi_top: Optional[IntelTLV]
+    cnvr_top: Optional[IntelTLV]
+    cnvi_bt: Optional[IntelTLV]
+    img_type: Optional[IntelTLV]
+    sbe_type: Optional[IntelTLV]
+
+
+def parse_intel_version_tlvs(data: bytearray) -> IntelVersionTlv:
+    tlvs = {
+        'cnvi_top': 0x10,
+        'cnvr_top': 0x11,
+        'cnvi_bt': 0x12,
+        'img_type': 0x1C,
+        'sbe_type': 0x2F,
+    }
+
+    parsed_tlvs: Dict[str, Optional[IntelTLV]] = {key: None for key in tlvs.keys()}
+    i = 0
+    while i < len(data):
+        try:
+            tlv_type = data[i]
+            tlv_length = data[i + 1]
+            tlv_value = data[i + 2 : i + 2 + tlv_length]
+
+            for name, type_code in tlvs.items():
+                if tlv_type == type_code:
+                    parsed_tlvs[name] = IntelTLV(tlv_type, tlv_length, tlv_value)
+                    break
+
+            i += 2 + tlv_length
+        except IndexError:
+            logging.error("TLV parse error")
+            continue
+
+    return IntelVersionTlv(
```

Comment:
suggestion: you could have all the fields default to None. Then in the while loop, just populate a dict for the properties you find, then invoke the constructor as `IntelVersionTlv(**tlvs)`